src/Phinx/Db/Adapter/MysqlAdapter.php

```diff
@@ -98,7 +99,59 @@ public function connect()
             }
         }
     }
-    
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasSchemaTable()
+    {
+        if (! $this->hasTable($this->getSchemaTableName())) {
+            return false;
+        }
+
+        if (! $this->hasIndex($this->getSchemaTableName(), 'version')) {
+            $this->updateSchemaTable();
+        }
+        return true;
+    }
+
+    protected function updateSchemaTable()
+    {
+        $this->getOutput()->writeln('<info>'.$this->getSchemaTableName() . ' schema out of date. updating schema</info>');
+        try {
+            $oldTable = $this->getSchemaTableName() . '_old';
+            $this->renameTable($this->getSchemaTableName(), $oldTable);
+            $this->createSchemaTable();
+            $this->execute('INSERT INTO ' . $this->getSchemaTableName() . ' SELECT version, start_time, end_time FROM ' . $oldTable . ' GROUP BY version');
+            $this->dropTable($oldTable);
+
+        } catch (\Exception $exception) {
+            throw new \InvalidArgumentException('There was a problem updating the schema table: ' . $exception->getMessage());
+        }
+        $this->getOutput()->writeln('<comment>finished updating '.$this->getSchemaTableName() . ' schema</comment>');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createSchemaTable()
```

Comment:
overrides original schema in PdoAdapter.
Adds version as a primary key (for `ON DUPLICATE KEY UPDATE`), and allows null `end_time`'s